src/cone_detector.cpp

```diff
@@ -111,19 +126,156 @@ bool ConeDetector::annotateCb(victoria_perception::AnnotateDetectorImage::Reques
+            if (point_hsv[2] > max_value) max_value = point_hsv[2];
+            image.at<cv::Vec3b>(cv::Point(point_number % width, point_number / width)) = cv::Vec3b(0, 0, 0);
+            pixels_in_cluster++;
+        }
+
+        point_number++;
+        label_first++;
+    }
+
+    std::cout << "There were " << pixels_in_cluster << " pixels in the cluster and the new ranges are "
+              << "min_hue: " << (int) min_hue << ", max_hue: " << (int) max_hue
+              << ", min_saturation: " << (int) min_saturation << ", max_saturation: " << (int) max_saturation
+              << ", min_value: " << (int) min_value << ", max_value: " << (int) max_value
+              << std::endl;
+
+    cv::namedWindow("KMEANS_annotated", cv::WINDOW_AUTOSIZE);
+    cv::imshow("KMEANS_annotated", image);
+    cv::waitKey(20);
+
+    // Sort the histogram, descending.
+    cv::Mat histSortIdx;
+    cv::sortIdx(cluster_histogram, histSortIdx, cv::SORT_EVERY_COLUMN + cv::SORT_DESCENDING);
+    std::cout << "cluster_histogram..." << std::endl << cluster_histogram << std::endl; //#####
+
+    // Print out the sort index of each cluster.
+    std::cout << "histSortIdx..." << std::endl << histSortIdx << std::endl; //#####
+
+    // Print out the number of label rows (i.e., number of pixels) and label columns (there is only one).
+    std::cout << "labels: " << labels.rows << " " << labels.cols << std::endl; //#####
+
+    for (int i = 0; i < histSortIdx.rows; i++) {
+        int sx = histSortIdx.at<int>(i);
+        cv::Point2f p = centers.at<cv::Point2f>(sx);
+        cv::Vec3b q = centers_u8.at<cv::Vec3b>(i);
+        std::cout << "sorted centers[sort index:" << i << " / cluster index:" << sx << "] histogram count: " << cluster_histogram.at<int>(sx) << ", point: " << p 
+                  << ", color: " << q << std::endl; //#####
+    }
+}
+
+bool ConeDetector::calibrateConeDetectionCb(victoria_perception::CalibrateConeDetection::Request &request,
+                                            victoria_perception::CalibrateConeDetection::Response &response) {
+    if (last_image_count_ <= 0) {
+        response.result = "No images received yet";
+        return false;
+    }
+
+    kmeansImage(last_image_);
+    response.result = "Calibrated";
+    return true;
+}
+
 bool ConeDetector::hullIsValid(std::vector<cv::Point>& hull) {
+    // Find the bounding rectangle so we can easily determine hull points which
+    // are above vs. below the horizontal centerline.
     cv::Rect bounding_rect = cv::boundingRect(cv::Mat(hull));
-    //float aspect_ratio = 1.0 * bounding_rect.width / bounding_rect.height;
-    float y_center = bounding_rect.y + (bounding_rect.height / 2.0);
+
+    float y_center = bounding_rect.y + (bounding_rect.height / 2.0); // The height of the horizontal centerline.
+
+    // Computer the upper/lower left/right x-coordinates of the hull. Like a bounding box only it's a 
+    // tight fit to the hull
     int top_left_x = 10000;
```

Comment:
Also, while there's no difference in the case of basic types, it would be better to use initialization rather than assignment, as with other types there will be a cost savings.